examples/petsc/bps.c

```diff
@@ -163,6 +164,11 @@ static PetscErrorCode RunWithDM(RunParams rp, DM dm,
   ierr = MatShellSetOperation(mat_O, MATOP_GET_DIAGONAL,
                               (void(*)(void))MatGetDiag); CHKERRQ(ierr);
   ierr = MatShellSetVecType(mat_O, vec_type); CHKERRQ(ierr);
+  if (rp->bp_choice > 1) {
+    ierr = MatNullSpaceCreate(rp->comm, PETSC_TRUE, 0, NULL, &null_space);
+    CHKERRQ(ierr);
+    ierr = MatSetNullSpace(mat_O, null_space); CHKERRQ(ierr);
+  }
```

Comment:
@jedbrown, I thought running a BP on a periodic mesh would be as simple as using `-dm_plex_box_bd periodic,periodic,periodic` and setting the null space to the constant vector, but I am getting `DIVERGED_INDEFINITE_MAT` (and `DIVERGED_INDEFINITE_PC` for bp1/2). Am I missing something obvious here?

Reply:
There used to be an issue in with tensor indices for periodic meshes, but @knepley has been hacking on this recently and it might have gone away. (I got part way through a somewhat tedious workaround in a branch that I think should be obsolete now.)

Reply:
It should be easy to check. Can you pull it down to Q1 and view the matrix?

Reply:
When I use Q1 and grab a a row out of the center of the assembled matrix with Dirichlet BCs, I get
```
row 32: (16, 2.18331e-16)  (17, -0.0333333)  (20, -0.0333333)  (21, -0.0166667)  (32, 0.533333)  (33, -2.00809e-17)  (36, 2.08309e-17)  (37, -0.0333333)  (48, 2.18696e-16)  (49, -0.0333333)  (52, -0.0333333)  (53, -0.0166667) 
```

But when I grab a row out of the matrix with periodic BCs, I get
```
row 0: (0, 0.0555556)  (1, -0.0555556)  (2, 0.0555556)  (3, -0.0555556)  (4, 0.0416667)  (5, -0.0555556)  (6, 0.0555556)  (7, -0.0555556)  (8, 0.0555556)  (9, -0.0555556)  (10, 0.0416667)  (11, -0.0555556)  (12, 0.0416667)  (13, -0.0277778)  (14, 0.0416667)  (15, -0.0555556)  (16, 0.0416667)  (17, -0.0555556)  (18, 0.0555556)  (19, -0.0555556)  (20, 0.0555556)  (21, -0.0555556)  (22, 0.0416667)  (23, -0.0555556)  (24, 0.0555556)  (25, -0.0555556)  (26, 0.0555556) 
```

So something looks strange to me

Reply:
Periodic has 27 nonzeros, which seems right for Q1 3D periodic hexes.

Reply:
Sure, but I'm used to the Laplacian  only having only one positive entry in a row and everything else negative

Reply:
If I had to guess, I would say you are using the coordinate field which jumps across Riemann sheets. When I calculate the geometry in Plex, I use the "localized" coordinates.

Reply:
Hmm, that could be... We use whatever comes out of DMGetCoordinatesLocal

Reply:
Both sets of coordinates are in there. The coordinates associated with vertices cannot express periodicity, so we also have a SG coordinate field stored on the cells.

Reply:
So is there a different set of coordinates I should be retrieving than what is in DMGetCoordinatesLocal?

Reply:
I'm surprised this doesn't generate an error in restriction setup due to mismatched closure sizes. This has been one of the longest standing disagreements between Matt and me. The "localized" coordinates are like a little DG space associated with the cell, but I don't think they're currently put into tensor ordering and I don't think it's easy to make that general. It also leads to singular operators if you have an ALE-type solve (unless you add Lagrange multipliers to enforce continuity elsewhere; much more complicated).

I have an old WIP to use the anchor system for periodicity, but it's now more than a year old. I have it on my plate to revive and finish in the next couple weeks, but Matt twisted my arm into being on a committee. :rofl:  https://gitlab.com/petsc/petsc/-/merge_requests/2178

Reply:
Ah, if I add 
```
ierr = DMLocalizeCoordinates(dm); CHKERRQ(ierr); /* needed for periodic */
```
then I get an error about sizes when creating the coordinate restriction.